packages/cli/cloud/src/link/action.ts

```diff
@@ -0,0 +1,204 @@
+            existingConfig.project.displayName
+              ? existingConfig.project.displayName
+              : existingConfig.project.name
+          )} is already linked to this local folder. Do you want to update the link?`,
+          default: false,
+        },
+      ]);
+
+      if (!shouldRelink) {
+        await trackEvent(ctx, cloudApiService, 'didNotLinkProject', {
+          currentProjectName: existingConfig.project?.name,
+        });
+        return null;
+      }
+    }
+  } catch (e) {
+    ctx.logger.debug('Failed to get project config', e);
+    ctx.logger.error('An error occurred while retrieving config data from your local project.');
+    return null;
+  }
+
+  return existingConfig;
+}
+
+async function getProjectsList(
+  ctx: CLIContext,
+  cloudApiService: CloudApiService,
+  existingConfig: LocalSave
+) {
+  const spinner = ctx.logger.spinner('Fetching your projects...\n').start();
+
+  try {
+    const {
+      data: { data: projectList },
+    } = await cloudApiService.listLinkProjects();
+    spinner.succeed();
+    const projects: ProjectsList = (projectList as unknown as Project[])
+      .filter(
+        (project: Project) =>
+          !(project.isMaintainer || project.name === existingConfig?.project?.name)
+      )
+      .map((project: Project) => {
+        return {
+          name: project.displayName,
+          value: { name: project.name, displayName: project.displayName },
+        };
+      });
+    if (projects.length === 0) {
+      ctx.logger.log("We couldn't find any projects available for linking in Strapi Cloud");
+      return null;
+    }
+    return projects;
+  } catch (e) {
+    spinner.fail('An error occurred while fetching your projects from Strapi Cloud.');
+    ctx.logger.debug('Failed to list projects', e);
+    return null;
+  }
+}
+
+async function getUserSelection(
+  ctx: CLIContext,
+  projects: ProjectsList
+): Promise<LinkProjectAnswer | null> {
+  const { logger } = ctx;
+  try {
+    const answer: LinkProjectInput = await inquirer.prompt([
+      {
+        type: 'list',
+        name: 'linkProject',
+        message: 'Which project do you want to link?',
+        choices: [...projects, { name: chalk.grey(`(${QUIT_OPTION})`), value: null }],
+      },
+    ]);
+
+    if (!answer.linkProject) {
+      return null;
+    }
+
+    return answer as LinkProjectAnswer;
+  } catch (e) {
+    logger.debug('Failed to get user input', e);
+    logger.error('An error occurred while trying to get your input.');
+    return null;
+  }
+}
+
+export default async (ctx: CLIContext) => {
+  const { getValidToken } = await tokenServiceFactory(ctx);
+  const token = await getValidToken(ctx, promptLogin);
+  const { logger } = ctx;
+
+  if (!token) {
+    return;
+  }
+
+  const cloudApiService = await cloudApiFactory(ctx, token);
+
+  const existingConfig: LocalSave | null = await getExistingConfig(ctx, cloudApiService);
+  // If user selects not to relink, return
+  if (!existingConfig) {
```

Comment:
I believe `getExistingConfig` function doesn't really do what it says, as it also query User to know if we should link or not.

Reply:
Well, I've asked that same question myself, but I think the purpose is to decompose the main function into subfunctions that make the code easier to read, maintain, and test. We can extract the query outside of it, but that would leave getExistingConfig as merely a wrapper for local.retrieve(). Other options include:

- Don’t abstract away getConfig and prompting; keep it in the main function.
- Create a function only for prompting for re-link, though it would be quite small to justify being outside.
- Rename `getExistingConfig` to something like `getConfigAndPrompt`.

Reply:
Resolved to refactor into single purpose functions